group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/assignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,427 @@
+                activeTasks = toCompactedTaskIds(activeTasksAssignments.get(memberId));
+            }
+            Map<String, Set<Integer>> standByTasks = new HashMap<>();
+
+            if (standbyTasksAssignments.containsKey(memberId)) {
+                standByTasks = toCompactedTaskIds(standbyTasksAssignments.get(memberId));
+            }
+            memberAssignments.put(memberId, new MemberAssignment(activeTasks, standByTasks, new HashMap<>()));
+        }
+
+        return new GroupAssignment(memberAssignments);
+    }
+
+    private Map<String, Set<Integer>> toCompactedTaskIds(final Set<TaskId> taskIds) {
+        Map<String, Set<Integer>> ret = new HashMap<>();
+        for (TaskId taskId : taskIds) {
+            ret.putIfAbsent(taskId.subtopologyId(), new HashSet<>());
+            ret.get(taskId.subtopologyId()).add(taskId.partition());
+        }
+        return ret;
+    }
+
+    private void assignActive(final Set<TaskId> activeTasks) {
+
+        // 1. re-assigning existing active tasks to clients that previously had the same active tasks
+        for (Iterator<TaskId> it = activeTasks.iterator(); it.hasNext();) {
+            final TaskId task = it.next();
+            final Member prevMember = activeTaskToPrevMember.get(task);
+            if (prevMember != null && hasUnfulfilledQuota(prevMember)) {
+                processIdToState.get(prevMember.processId).addTask(prevMember.memberId, task, true);
+                updateHelpers(prevMember, task, true);
+                it.remove();
+            }
+        }
+
+        // 2. re-assigning tasks to clients that previously have seen the same task (as standby task)
+        for (Iterator<TaskId> it = activeTasks.iterator(); it.hasNext();) {
+            final TaskId task = it.next();
+            final Set<Member> prevMembers = standbyTaskToPrevMember.get(task);
+            final Member prevMember = findMemberWithLeastLoad(prevMembers, task, true);
+            if (prevMember != null && hasUnfulfilledQuota(prevMember)) {
+                processIdToState.get(prevMember.processId).addTask(prevMember.memberId, task, true);
+                updateHelpers(prevMember, task, true);
+                it.remove();
+            }
+        }
+
+        // 3. assign any remaining unassigned tasks
+        for (Iterator<TaskId> it = activeTasks.iterator(); it.hasNext();) {
+            final TaskId task = it.next();
+            final Set<Member> allMembers = processIdToState.entrySet().stream().flatMap(entry -> entry.getValue().memberToTaskCounts().keySet().stream()
+                .map(memberId -> new Member(entry.getKey(), memberId))).collect(Collectors.toSet());
+            final Member member = findMemberWithLeastLoad(allMembers, task, false);
+            if (member == null) {
+                log.error("Unable to assign active task {} to any member.", task);
+                throw new TaskAssignorException("No member available to assign active task {}." + task);
```

Comment:
In which cases can this happen, are we sure we want to crash in all of those?

For example, I guess this can happen if we have no members in the group. If we are going to crash in the case where no members are in the group, it would be good to document this as a contract in the task assignor interface.

Are there cases where there are members in the group and I cannot find an assignment?

Reply:
>this can happen if we have no members in the group.

Yes, back to `findMemberWithLeastLoad`, it is the only case.
The current `StickyTaskAssignor` did not consider this case ([here](https://github.com/apache/kafka/blob/04567cdb222badafaf1ff77407800c4b3acc550b/streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java#L182-L186))

Reply:
Yeah, I think the current StickytaskAssignor does not take this case into account, because it's a member computing the assignment, so by this we know there must be at least one member. For a member to consider its own nonexistence would be quite philosophical :P

So it seems to me, we either need to make assignors work for empty groups (which is probably just logging something and returning an empty assignment), or we document in the interface that assignors will throw `TaskAssignorException` for empty groups.